internal/metrics/metrics.go

```diff
@@ -1,17 +1,224 @@
+// registers these metrics to the prometheus registry, and returns the registry and the metrics.
+// The registry and metrics can be modified from this struct from anywhere in the codebase.
+func newPrometheusMetrics() *PrometheusMetrics {
+	m := &PrometheusMetrics{
+		registry:                       prometheus.NewRegistry(),
+		CertificateRequests:            certificateRequestsMetric(),
+		OutstandingCertificateRequests: outstandingCertificateRequestsMetric(),
+		Certificates:                   certificatesMetric(),
+		ExpiredCertificates:            expiredCertificatesMetric(),
+		CertificatesExpiringIn1Day:     certificatesExpiringIn1DayMetric(),
+		CertificatesExpiringIn7Days:    certificatesExpiringIn7DaysMetric(),
+		CertificatesExpiringIn30Days:   certificatesExpiringIn30DaysMetric(),
+		CertificatesExpiringIn90Days:   certificatesExpiringIn90DaysMetric(),
+
+		RequestsTotal:    requestsTotalMetric(),
+		RequestsDuration: requestDurationMetric(),
+	}
+	m.registry.MustRegister(m.CertificateRequests)
+	m.registry.MustRegister(m.OutstandingCertificateRequests)
+	m.registry.MustRegister(m.Certificates)
+	m.registry.MustRegister(m.ExpiredCertificates)
+	m.registry.MustRegister(m.CertificatesExpiringIn1Day)
+	m.registry.MustRegister(m.CertificatesExpiringIn7Days)
+	m.registry.MustRegister(m.CertificatesExpiringIn30Days)
+	m.registry.MustRegister(m.CertificatesExpiringIn90Days)
+
+	m.registry.MustRegister(m.RequestsTotal)
+	m.registry.MustRegister(m.RequestsDuration)
+
+	m.registry.MustRegister(collectors.NewGoCollector())
+	m.registry.MustRegister(collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}))
+	return m
+}
+
+// GenerateMetrics receives the live list of csrs to calculate the most recent values for the metrics
+// defined for prometheus
+func (pm *PrometheusMetrics) GenerateMetrics(csrs []certdb.CertificateRequest) {
+	var csrCount float64 = float64(len(csrs))
+	var outstandingCSRCount float64
+	var certCount float64
+	var expiredCertCount float64
+	var expiringIn1DayCertCount float64
+	var expiringIn7DaysCertCount float64
+	var expiringIn30DaysCertCount float64
+	var expiringIn90DaysCertCount float64
+	for _, entry := range csrs {
+		if entry.Certificate == "" {
+			outstandingCSRCount += 1
+			continue
+		}
+		if entry.Certificate == "rejected" {
+			continue
+		}
+		certCount += 1
+		expiryDate := certificateExpiryDate(entry.Certificate)
+		daysRemaining := time.Until(expiryDate).Hours() / 24
+		if daysRemaining < 0 {
+			expiredCertCount += 1
+		} else {
+			if daysRemaining < 1 {
+				expiringIn1DayCertCount += 1
+			}
+			if daysRemaining < 7 {
+				expiringIn7DaysCertCount += 1
+			}
+			if daysRemaining < 30 {
+				expiringIn30DaysCertCount += 1
+			}
+			if daysRemaining < 90 {
+				expiringIn90DaysCertCount += 1
+			}
+		}
+	}
+	pm.CertificateRequests.Set(csrCount)
+	pm.OutstandingCertificateRequests.Set(outstandingCSRCount)
+	pm.Certificates.Set(certCount)
```

Comment:
This not a review comment but a question to learn :D
What does this Set do? what is the end result?
Where does the value fit in the following?
```
func certificatesMetric() prometheus.Gauge {
	metric := prometheus.NewGauge(prometheus.GaugeOpts{
		Name: "certificates",
		Help: "Total number of certificates provided to certificate requests",
	})
	return metric
}
```

Reply:
The name `gauge` comes from a real life [gauge](https://catalog.noshok.com/Asset/60_800_200_PSI-3_4.jpg), It's simply one of three types of metrics that can be collected.

A prom [Gauge](https://prometheus.io/docs/concepts/metric_types/#gauge) is initialized with options that describe what it is, but it is fundamentally a gauge with a discrete number as a value. We can Increment or Decrement this value by 1, but the prom golang client [recommends](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#NewGauge) that you just Set the value directly if you can.